File: go/PartitionLL/PartitionLL.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time -> 0(N) Space -> 0(1)
func partition(head *ListNode, x int) *ListNode {
	curr := head
	var lessHead, less, prev, pivot *ListNode
	for curr != nil {
		if curr.Val >= x {
			pivot = curr
			currPrev := prev
			for curr != nil {
				if curr.Val < x {
					if lessHead == nil {
						less = curr
						lessHead = less
					} else {
						less.Next = curr
						less = less.Next
					}
					currPrev.Next = curr.Next
				} else {
					currPrev = curr
				}
				curr = curr.Next
			}
			break
		}
		prev = curr
		curr = curr.Next
	}

	if less != nil {
		if prev == nil {
			less.Next = head
			return lessHead
		}

		prev.Next = lessHead
		less.Next = pivot
	}
	return head
}

func main() {
	head := &ListNode{}
	currHead := head
	for _, val := range []int{2, 3, 4, 3, 1, 2, 5} {
		currHead.Val = val
		currHead.Next = &ListNode{}
		currHead = currHead.Next
	}

	res := partition(head, 3)
	var array []int
	for res != nil {
		array = append(array, res.Val)
		res = res.Next
	}
	fmt.Println(array)
}

File: python/partitionll.py
class ListNode:
    def __init__(self, val=0, next=None) -> None:
        self.val = val
        self.next = next


# Time -> 0(N) Space -> 0(1)
def partition(head: ListNode | None, x: int) -> ListNode | None:
    curr = head
    lessHead, less, prev, pivot = None, None, None, None
    while curr is not None:
        if curr.val >= x:
            pivot = curr
            currPrev = prev
            while curr is not None:
                if curr.val < x:
                    if less is None:
                        less = curr
                        lessHead = less
                    else:
                        less.next = curr
                        less = less.next

                    if currPrev is not None:
                        currPrev.next = curr.next

                else:
                    currPrev = curr
                curr = curr.next
            break

        prev = curr
        curr = curr.next

    if less is not None:
        if prev is None:
            less.next = head
            return lessHead

        prev.next = lessHead
        less.next = pivot

    return head


head = ListNode()
currHead = head
for val in [2, 3, 4, 3, 1, 2, 5]:
    currHead.val = val
    currHead.next = ListNode()
    currHead = currHead.next

res = partition(head, 3)

array = []
while res is not None:
    array.append(res.val)
    res = res.next

print(array)
